addons/udes_stock/models/stock_picking.py

```diff
@@ -739,3 +740,268 @@ def raise_stock_investigation(self, reason, quants, location):
         # Try to re-assign the picking after creating the stock investigation
         # picking as we have reserved the problematic stock
         self.action_assign()
+
+    def construct_package_hierarchy_links(self):
+        # TODO: add link/unlink to parent if needed like old _set_u_result_parent_package_id()
+        # PackageHierarchyLink = self.env["package.hierarchy.link"]
+        # new_link = PackageHierarchyLink.create(
+        #     {
+        #         "child_id": result_package.id,
+        #         "parent_id": result_parent_package.parent_id.id,
+        #         "move_line_ids": [(6, 0, mls.ids)],
+        #     }
+        # )
+        super().construct_package_hierarchy_links()
+
+    def _check_entire_pack(self):
+        """
+        Override to avoid values at result_package_id when
+        user scans products/avoid _check_entire_pack of products
+        """
+        pickings = self.filtered(
+            lambda p: p.picking_type_id.u_user_scans != "product"
+            and p.picking_type_id.u_target_storage_format != "product"
+        )
+        super(StockPicking, pickings)._check_entire_pack()
+
+    def log_name(self):
+        """This can probably go to models later"""
+        return f"Picking({self.id, self.name}"
+
+    def assert_valid_state(self):
+        """Checks if the transfer is in a valid state, i.e., not done or cancel
+        otherwise it raises and error
+        """
+        self.ensure_one()
+        if self.state in ("done", "cancel"):
+            raise ValidationError(_("Wrong state %s for %s") % (self.state, self.log_name()))
+
+    def _action_done(self):
+        """Extend to add logging"""
+        with self.statistics() as stats:
+            res = super()._action_done()
+        _logger.info(
+            "%s udes_stock::_action_done in %.2fs, %d queries", self.log_name(), stats.elapsed, stats.count
+        )
+        return res
+
+    def add_unexpected_parts(self, product_quantities):
+        """ TODO not needed for now, add to readme when implemented
+        """
+        pass
+
+    def _prepare_result_packages(
+        self,
+        product_ids,
+        package,
+        result_package_name,
+        result_parent_package_name,
+        target_storage_format=None,
+    ):
+        """
+        Compute result_package and result_parent_package based on the u_target_storage_format of the
+        picking type (if `target_storage_format` is not supplied) + the input parameters.
+        """
+        Package = self.env["stock.quant.package"]
+
+        result_parent_package = None
+
+        if target_storage_format is None:
+            target_storage_format = self.picking_type_id.u_target_storage_format
+        scan_parent_package_end = self.picking_type_id.u_scan_parent_package_end
+
+        if target_storage_format == "pallet_packages":
+
+            # CASE A: both package names given
+            if result_package_name and result_parent_package_name:
```

Comment:
We could with newer verions for storage format configs use cases, and then have this consistent throughout but not sure the python version used so feel free to ignore.

Reply:
At the moment we have ported the same funciton we have in 11. But in future work we could have a class, like in suggested locations, that handles what should happen for a given target storage format.